### PR TITLE
Add hook method to track performance of separate build stages

### PR DESCRIPTION
**Pull request by samogot**
Track performance of different builder stages in build_web_compiler

### CODE REVIEW

File: build/lib/src/builder/build_step.dart [hunk 1]
```diff
@@ -63,4 +63,26 @@ abstract class BuildStep implements AssetReader, AssetWriter {
...
+  ///
+  /// If performance tracking is enabled, tracks [action] as separate stage
+  /// identified by [label]. Otherwise just runs [action].
+  ///
+  /// Returns value returned by [action].
+  /// [action] can be async function returning [Future].
+  T trackStage<T>(String label, T Function() action);
+}
+
+abstract class StageTracker {
```
Comment (samogot) [inline build/lib/src/builder/build_step.dart]: I'm not sure this is the best place to add this

File: build/pubspec.yaml [hunk 2]
```diff
@@ -17,9 +17,18 @@ dependencies:
...
 dev_dependencies:
-  build_resolvers: ^0.2.1
+  build_resolvers: ^0.2.2+4
   build_runner: ^0.9.0
   build_test: ^0.10.0
   build_vm_compilers: ^0.1.0
   pedantic: ^1.0.0
   test: ^1.2.0
+
+# Remove when release
```
Comment (jakemac53) [inline build/pubspec.yaml]: nit: we don't need this comment (you can't publish with dependency overrides)

File: build/lib/src/builder/build_step_impl.dart [hunk 3]
```diff
@@ -55,8 +57,15 @@ class BuildStepImpl implements BuildStep {
...
-      this._writer, this._rootPackage, this._resolvers, this._resourceManager)
+  BuildStepImpl(
+      this.inputId,
+      Iterable<AssetId> expectedOutputs,
+      this._reader,
+      this._writer,
+      this._rootPackage,
+      this._resolvers,
+      this._resourceManager,
+      this._stageTracker)
```
Comment (jakemac53) [inline build/lib/src/builder/build_step_impl.dart]: nit: make this optional and use the NoOpStageTracker by default

File: build_runner_core/lib/src/generate/build_impl.dart [hunk 4]
```diff
@@ -825,3 +828,13 @@ String _actionLoggerName(
       : primaryInput.uri;
   return '${phase.builderLabel} on $asset';
 }
+
+class BuildStepStageTracker implements StageTracker {
+  final BuilderActionTracker _tracker;
```
Comment (jakemac53) [inline build_runner_core/lib/src/generate/build_impl.dart]: We should just make this class implement StageTracker imo - that way we don't have to create a bunch of these wrapper objects

File: build_web_compilers/lib/src/dev_compiler_builder.dart [hunk 5]
```diff
@@ -63,7 +63,8 @@ class DevCompilerBuilder implements Builder {
 Future _createDevCompilerModule(
     Module module, BuildStep buildStep, bool useKernel,
     {bool debugMode = true}) async {
-  var transitiveDeps = await module.computeTransitiveDependencies(buildStep);
+  var transitiveDeps = await buildStep.trackStage('TransitiveDependencies',
```
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_builder.dart]: nit: `CollectTransitiveDeps`

File: build_web_compilers/lib/src/dev_compiler_builder.dart [hunk 6]
```diff
@@ -168,7 +170,8 @@ Future _createDevCompilerModule(
     var driverResource =
         useKernel ? dartdevkDriverResource : dartdevcDriverResource;
     var driver = await buildStep.fetchResource(driverResource);
-    response = await driver.doWork(request);
+    response =
+        await buildStep.trackStage('DoWork', () => driver.doWork(request));
```
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_builder.dart]: nit: `Compile`

File: build/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,7 @@
+## 0.12.8
+
+- Add hook method to track performance of separate build stages.
```
Comment (jakemac53) [inline build/CHANGELOG.md]: Lets reference the actual method so this is more useful:
```
- Added the `T trackStage<T>(String label, T Function() action);` method to `BuildStep`.
```

**Comment by jakemac53** [comment]
> I'd like to get some verification that some builder authors want to use this before we commit.
> 
> I think @jakemac53 has wanted it for at least one of the builders we authored, but we might also want to check with Angular and see if they think it's useful

I think even just our own use justifies this (see the build_web_compilers updates in this pr). It helps a lot for tracking the individual portions of your builder.

File: build/lib/src/builder/build_step.dart [hunk 8]
```diff
@@ -69,20 +69,29 @@ abstract class BuildStep implements AssetReader, AssetWriter {
...
+  /// be tracked as single time slice from the beginning of the stage till
+  /// completion of Future returned by [action].
+  ///
+  /// Otherwise all separate time slices of asynchronous execution will be
+  /// tracked, but waiting for external resources will be a gap.
+  ///
   /// Returns value returned by [action].
   /// [action] can be async function returning [Future].
-  T trackStage<T>(String label, T Function() action);
+  T trackStage<T>(String label, T Function() action, {bool external = false});
```
Comment (natebosch) [inline build/lib/src/builder/build_step.dart]: how about `isExternal`

**Comment by samogot** [comment]
@jakemac53 @natebosch ptal :)

**Pull request merged**